src/prefect/server/database/configurations.py

```diff
@@ -246,32 +266,76 @@ async def dispose_engine(cache_key):
 
         await add_event_loop_shutdown_callback(partial(dispose_engine, cache_key))
 
-    def setup_sqlite(self, conn, named=True):
+    def setup_sqlite(self, conn, record):
         """Issue PRAGMA statements to SQLITE on connect. PRAGMAs only last for the
         duration of the connection. See https://www.sqlite.org/pragma.html for more info.
         """
+        # workaround sqlite transaction behavior
+        self.begin_sqlite_conn(conn, record)
+
+        cursor = conn.cursor()
+
+        # write to a write-ahead-log instead and regularly commit the changes
+        # this allows multiple concurrent readers even during a write transaction
+        # even with the WAL we can get busy errors if we have transactions that:
+        #  - t1 reads from a database
+        #  - t2 inserts to the database
+        #  - t1 tries to insert to the database
+        # this can be resolved by using the IMMEDIATE transaction mode in t1
+        cursor.execute("PRAGMA journal_mode = WAL;")
+
         # enable foreign keys
-        conn.execute(sa.text("PRAGMA foreign_keys = ON;"))
+        cursor.execute("PRAGMA foreign_keys = ON;")
 
-        # write to a write-ahead-log instead and regularly
-        # commit the changes
-        # this allows multiple concurrent readers even during
-        # a write transaction
-        conn.execute(sa.text("PRAGMA journal_mode = WAL;"))
+        # disable legacy alter table behavior as it will cause problems during
+        # migrations when tables are renamed as references would otherwise be retained
+        # in some locations
+        # https://www.sqlite.org/pragma.html#pragma_legacy_alter_table
+        cursor.execute("PRAGMA legacy_alter_table=OFF")
```

Comment:
It's kind of wild, but this was ON by default and causing errors during migrations. I'm not entirely sure why it was not a problem before.